users/src/main/scala/users/services/usermanagement/interpreters.scala

```diff
@@ -0,0 +1,238 @@
+package users.services.usermanagement
+
+import java.util.UUID
+import java.time.OffsetDateTime
+
+import cats.data.EitherT
+import cats.implicits._
+import users.config._
+import users.domain._
+import users.persistence.repositories._
+
+import scala.concurrent._
+import scala.util.Random
+
+object Interpreters {
+  def default(
+      userRepository: UserRepository
+  )(
+      implicit
+      ec: ExecutionContext
+  ): Algebra[Future[?]] =
+    new DefaultInterpreter(userRepository)
+
+  def unreliable(
+      underlying: Algebra[Future[?]],
+      config: ServicesConfig.UsersConfig
+  )(
+      implicit
+      ec: ExecutionContext
+  ): Algebra[Future[?]] =
+    new UnreliableInterpreter(underlying, config)
+}
+
+final class DefaultInterpreter private[usermanagement] (
+    userRepository: UserRepository
+)(
+    implicit
+    ec: ExecutionContext
+) extends Algebra[Future[?]] {
+  import User._
+
+  def generateId(): Future[Id] =
+    Future.successful(Id(UUID.randomUUID().toString))
+
+  def get(
+      id: Id
+  ): Future[Error Either User] =
+    for {
+      maybeUser ← userRepository.get(id)
+      result = maybeUser.toRight(Error.NotFound)
+    } yield result
+
+  def signUp(
+      userName: UserName,
+      emailAddress: EmailAddress,
+      password: Option[Password]
+  ): Future[Error Either User] =
+    (for {
+      maybeUser ← EitherT[Future, Error, Option[User]] {
+        userRepository.getByUserName(userName).map(Right.apply)
+      }
+      id ← EitherT[Future, Error, Id](generateId().map(Right.apply))
+      result ← EitherT.fromEither[Future] {
+        if (maybeUser.nonEmpty) Left(Error.AlreadyExists: Error)
+        else
+          Right(User(id, userName, emailAddress, password, OffsetDateTime.now()))
```

Comment:
Think you have to do `repository.insert` or `save` as well?

Reply:
good catch